### PR TITLE
Add kamailio Integration

### PR DESCRIPTION
**Pull request by devopsec**
### What does this PR do?

Adds integration for kamailio SIP server.

### Motivation

No integrations for this service existed.

### Review checklist

- [x] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [x] Feature or bugfix has tests
- [x] Git history is clean
- [x] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### Additional Notes

None.

### CODE REVIEW

**Comment by devopsec** [comment]
fixed CI build integration

**Comment by buraizu** [comment]
Thanks, created DOCS-4458 to review

**Review by buraizu — CHANGES_REQUESTED** [review]

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -67,6 +67,7 @@
 /ilert/                                  @yacut support@ilert.com
 /insightfinder/                          @dearmore @erinlmcmahon support@insightfinder.com
 /rum_javascript/                         @DataDog/rum-app
+/kamailio/                               @devopsec
```
Comment (FlorentClarret) [inline .github/CODEOWNERS:70]: Could you also add `kamailio` to [this section](https://github.com/DataDog/integrations-extras/blob/master/.github/CODEOWNERS#L169) so our doc team will be notified when needed?

File: kamailio/assets/service_checks.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
+[
+  {
+    "agent_version": "7.0.0",
```
Comment (FlorentClarret) [inline kamailio/assets/service_checks.json:3]:
```suggestion
    "agent_version": "7.40.1",
```

File: kamailio/assets/service_checks.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    "integration": "kamailio",
+    "check": "kamailio.services_up",
+    "statuses": ["ok", "warning", "critical"],
+    "groups": [],
+    "name": "Kamailio Services Up",
+    "description": "Returns `CRITICAL` if the services are down, `WARNING` if the agent is misconfigured, or `OK` if the services are running."
+  },
+  {
+    "agent_version": "7.0.0",
```
Comment (FlorentClarret) [inline kamailio/assets/service_checks.json:12]:
```suggestion
    "agent_version": "7.40.1",
```

File: kamailio/datadog_checks/kamailio/__about__.py [hunk 4]
```diff
@@ -0,0 +1 @@
+__version__ = '0.0.1'
```
Comment (FlorentClarret) [inline kamailio/datadog_checks/kamailio/__about__.py:1]:
```suggestion
__version__ = '1.0.0'
```

File: kamailio/datadog_checks/kamailio/check.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        :rtype:             None
+        """
+
+        # validate instance configs
+        jsonrpc_api_url = instance.get('jsonrpc_api_url')
+        if jsonrpc_api_url is None:
+            self.service_check(
+                "services_up", AgentCheck.WARNING, message='agent misconfiguration: jsonrpc_api_url is not valid'
+            )
+            return None
```
Comment (FlorentClarret) [inline kamailio/datadog_checks/kamailio/check.py:33]: See https://github.com/DataDog/integrations-extras/pull/1610/files#r1031302293

File: kamailio/manifest.json [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    "description": "Monitoring agent for kamailio SIP server",
+    "title": "kamailio",
+    "media": [],
+    "classifier_tags": [
+      "Supported OS::Linux",
+      "Supported OS::Windows",
+      "Supported OS::macOS"
+    ]
+  },
+  "assets": {
```
Comment (FlorentClarret) [inline kamailio/manifest.json:20]: You should add your dashboard here

File: kamailio/setup.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    # License
+    license='BSD-3-Clause',
+    # See https://pypi.org/classifiers
+    classifiers=[
+        'Development Status :: 5 - Production/Stable',
+        'Intended Audience :: Developers',
+        'Intended Audience :: System Administrators',
+        'Topic :: System :: Monitoring',
+        'License :: OSI Approved :: BSD License',
+        'Programming Language :: Python :: 2.7',
```
Comment (FlorentClarret) [inline kamailio/setup.py:66]:
```suggestion
```

**Review by FlorentClarret — CHANGES_REQUESTED** [review]
Hi again! thanks for this PR.

I left some comments but please also take a look at the ones I added on your Homer PR and apply them to this one too if appropriate. 

Thanks!

File: kamailio/manifest.json (hunk 6 above)
Comment (devopsec) [reply kamailio/manifest.json:32]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/setup.py (hunk 7 above)
Comment (devopsec) [reply kamailio/setup.py]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/datadog_checks/kamailio/check.py (hunk 5 above)
Comment (devopsec) [reply kamailio/datadog_checks/kamailio/check.py]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/datadog_checks/kamailio/__about__.py (hunk 4 above)
Comment (devopsec) [reply kamailio/datadog_checks/kamailio/__about__.py]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/assets/service_checks.json (hunk 3 above)
Comment (devopsec) [reply kamailio/assets/service_checks.json]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/assets/service_checks.json (hunk 2 above)
Comment (devopsec) [reply kamailio/assets/service_checks.json]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: .github/CODEOWNERS (hunk 1 above)
Comment (devopsec) [reply .github/CODEOWNERS:70]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/README.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+## Setup
+
+### Building
+
+1. Install the [Datadog Developer Toolkit][3].
+
+2. Clone the `integrations-extras` repository:
+
+```shell
+git clone https://github.com/DataDog/integrations-extras.git.
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/README.md]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/manifest.json [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  "app_uuid": "8ed5bcf6-21cb-4eae-86ff-11249b4212f2",
+  "app_id": "kamailio",
+  "display_on_public_website": false,
+  "tile": {
+    "overview": "README.md#Overview",
+    "configuration": "README.md#Setup",
+    "support": "README.md#Support",
+    "changelog": "CHANGELOG.md",
+    "description": "Monitoring agent for kamailio SIP server",
+    "title": "kamailio",
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/manifest.json]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/README.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+### Service Checks
+
+1. services_up - validates the kamailio services are running properly
+2. metrics_up  - validates the metrics were sent to datadog successfully
+
+For more information see [service_checks.json][10] for the service check definitions. 
+
+### Events
+
+The kamailio agent does not include any events.
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/README.md]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/README.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+```
+
+For more information see [metadata.csv][11] for an example of metrics that may be collected.
+
+### Service Checks
+
+1. services_up - validates the kamailio services are running properly
+2. metrics_up  - validates the metrics were sent to datadog successfully
+
+For more information see [service_checks.json][10] for the service check definitions.
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/README.md]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/README.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+```shell
+ddev -e release build kamailio
+```
+
+### Installing
+
+Once you built the package wheel, install it on a host:
+
+1. If the datadog agent is not installed yet [Install the Datadog Agent][2].
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/README.md]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/README.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+4. Build the `datadog-kamailio` package:
+
+```shell
+ddev -e release build kamailio
+```
+
+### Installing
+
+Once you built the package wheel, install it on a host:
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/README.md]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/README.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+ddev config set extras $(pwd)
+```
+
+4. Build the `datadog-kamailio` package:
+
+```shell
+ddev -e release build kamailio
+```
+
+### Installing
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/README.md]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/README.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
+# Agent Check: kamailio
+
+## Overview
+
+This check monitors [kamailio][1].
+
+## Setup
+
+### Building
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/README.md]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/README.md [hunk 16]
```diff
@@ -0,0 +1,96 @@
+# Agent Check: kamailio
+
+## Overview
+
+This check monitors [kamailio][1].
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/README.md]: resolved in cf85a7a2cd46cc4e6c399e639ff477102a24aead

File: kamailio/README.md [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+[2]: https://app.datadoghq.com/account/settings#agent
+[3]: https://docs.datadoghq.com/developers/integrations/new_check_howto/#developer-toolkit
+[4]: https://dopensource.com/
+[5]: https://kamailio.org/docs/modules/5.5.x/
+[6]: https://www.kamailio.org/wiki/cookbooks/5.5.x/core#statistics
+[7]: https://github.com/DataDog/integrations-extras/blob/master/kamailio/datadog_checks/kamailio/data/conf.yaml.example
+[8]: https://docs.datadoghq.com/agent/guide/agent-commands/#start-stop-and-restart-the-agent
+[9]: https://docs.datadoghq.com/agent/guide/agent-commands/#agent-status-and-information
+[10]: https://github.com/DataDog/integrations-extras/blob/master/kamailio/assets/service_checks.json
+[11]: https://github.com/DataDog/integrations-extras/blob/master/kamailio/metadata.csv
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/README.md]: resolved in 9b557b755e7c8cd700baac78164cd96643818001

File: kamailio/README.md [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+1. Edit the config file `/etc/datadog-agent/conf.d/kamailio.d/conf.yaml.example` per your settings.  
+   An example config can be found [here][7].  
+   Save the modified file as `/etc/datadog-agent/conf.d/kamailio.d/conf.yaml`.
+
+2. [Restart the Agent][8].
+
+### Validation
+
+1. Check the [agent status][9] and make sure `kamailio` is listed under the Checks section.
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/README.md]: resolved in resolved in 9b557b755e7c8cd700baac78164cd96643818001

File: kamailio/README.md [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+
+Currently all statistics produced from kamailio are forwarded to datadog.  
+See the [kamailio modules][5] documentation for information on statistics produced by each module.    
+To see what statistics your kamailio instance will produce run:
+
+```shell
+kamcmd stats.get_statistics all
+```
+
+For more information see [metadata.csv][11] for an example of metrics that may be collected.
```
[the remark this answers is not in the document]
Comment (devopsec) [reply kamailio/README.md]: resolved in 9b557b755e7c8cd700baac78164cd96643818001

**Comment by devopsec** [comment]
@buraizu it seems that serverside the docker instance for Kamailio is not spinning up, locally I am not having this issue.  
I also attached an example [test.log][1] of running the tests locally.

[1]: https://github.com/DataDog/integrations-extras/files/10435208/test.log

**Comment by FlorentClarret** [comment]
Hi @devopsec 👋 One problem is the tox file you are using. Currently, only py38 is supported. You'll need to remove the other environments with different python versions. About the `pre_command` you defined, I don't think this will be needed.

Also please note that we are not using tox anymore, we moved to [hatch](https://hatch.pypa.io/latest/). Could you modify the config to use hatch instead? The `ddev create` command should now generate an hatch file and not a tox one. You could also have a look at the other integrations defined in the repo if need be. Please, let me know if you need anything!

**Comment by devopsec** [comment]
@FlorentClarret Tox was working in the CI build though, I can try migrating over to hatch but I assume it will have the same issue with the container either way. Is there a way to manually inspect the container from the CI build? My hunch is that the `ddev-kamailio` container is spinning up but the process is not running correctly in that environment (compared to testing on my local environment).

About the `pre_command`, I documented the reason in the tox file:

```
    # workaround moving docker context or symlinked files by copying the check data files for docker
    # this also avoids the alternative situation of building the wheel with data files in an external location
```

The alternate is symlinks or moving the docker context to the integration root directory; due to the data files being included in the wheel, both of which are undesirable. If you have an alternative approach, I am all ears!

File: kamailio/README.md [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+   This setting is also located in the `kamailio.d/conf.yaml.example` configuration file.  
+   You wil also need to allow the `dd-agent` user to read the loaded modules from memory.  
+   Another pre-configured *sudoers* config bundled can handle this.  
+   To do so run the following command:
+
+```shell
+cp /opt/datadog-agent/embedded/lib/python*/site-packages/datadog_checks/kamailio/sudoers.d/dd-agent-procmaps /etc/sudoers.d/
+```
+
+3. Now you can edit any optional parameters in `kamailio.d/conf.yaml.example` to suit your environement.
```
Comment (buraizu) [inline kamailio/README.md:59]:
```suggestion
3. Edit any optional parameters in `kamailio.d/conf.yaml.example` to suit your environement.  
```

File: kamailio/README.md [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+
+### Validation
+
+Run the [Agent's `status` subcommand][9] and look for `kamailio` under the Running Checks section.
+
+```shell
+sudo -u dd-agent datadog-agent status
+```
+
+Here is an example of what the output should look like when the installed succeeded:
```
Comment (buraizu) [inline kamailio/README.md:73]:
```suggestion
Here is an example of what the output should look like when the installation succeeds: 
```

File: kamailio/README.md [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+      Service Checks: Last Run: 2, Total: 282
+      Average Execution Time : 19ms
+      Last Execution Date : 2023-01-12 23:10:48 EST / 2023-01-13 04:10:48 UTC (1673583048000)
+      Last Successful Execution Date : 2023-01-12 23:10:48 EST / 2023-01-13 04:10:48 UTC (1673583048000)
+```
+
+## Data Collected
+
+The data collected depends on what modules are loaded in your Kamailio configuration.  
+The intent is that all statistics that can help you better manage your system are forwarded for processing.
```
Comment (buraizu) [inline kamailio/README.md:95]:
```suggestion
The data collected depends on the modules loaded in your Kamailio configuration.  
Forward all statistics that can help you better manage your system for processing.
```

**Review by buraizu — CHANGES_REQUESTED** [review]
Noticed some additional issues and suggested unifying the metric description formatting.

**Comment by alopezz** [comment]
Closing this due to inactivity. Feel free to re-open if you start working on it again.

**Pull request closed without merging**